features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/service/DefaultTopologyService.java

```diff
@@ -0,0 +1,239 @@
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Throwables;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Lists;
+
+/**
+ * @author mvrueden
+ */
+public class DefaultTopologyService implements TopologyService {
+
+    /**
+     * Key class for the GraphProviderCache.
+     */
+    private static class GraphProviderKey {
+        private final String metaTopologyId;
+        private final String namespace;
+
+        public GraphProviderKey(String metaTopologyId, String namespace) {
+            this.metaTopologyId = metaTopologyId;
+            this.namespace = namespace;
+        }
+
+        public String getMetaTopologyId() {
+            return metaTopologyId;
+        }
+
+        public String getNamespace() {
+            return namespace;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(metaTopologyId, namespace);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || !(o instanceof GraphProviderKey)) {
+                return false;
+            }
+            final GraphProviderKey key = (GraphProviderKey) o;
+            return Objects.equals(metaTopologyId, key.metaTopologyId)
+                    && Objects.equals(namespace, key.namespace);
+        }
+    }
+
+    private static final LayoutAlgorithm DEFAULT_LAYOUT_ALGORITHM = new D3TopoLayoutAlgorithm();
+
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultTopologyService.class);
+
+    private final List<MetaTopologyProvider> providers = Lists.newArrayList();
+
+    private final LoadingCache<GraphProviderKey, GraphProvider> graphProviderCache;
+
+    private BundleContext bundleContext;
+
+    public DefaultTopologyService() {
+        this(30);
+    }
+
+    /**
+     * @param cacheTimeout The cache timeout in seconds.
+     *                     0 disables caching.
+     */
+    public DefaultTopologyService(long cacheTimeout) {
+        graphProviderCache = CacheBuilder
+                .newBuilder()
+                .expireAfterWrite(Math.max(0, cacheTimeout), TimeUnit.SECONDS)
+                // Due to graphml we may end up having a ton of graph providers.
+                // Caching is limited to 50, which is still plenty
+                .maximumSize(50)
+                .build(new CacheLoader<GraphProviderKey, GraphProvider>() {
+                    @Override
+                    public GraphProvider load(GraphProviderKey key) throws Exception {
+                        final MetaTopologyProvider metaTopologyProvider = getMetaTopologyProvider(key.getMetaTopologyId());
+                        metaTopologyProvider.reload(key.getNamespace());
```

Comment:
How do we handle concurrency now that the MetaTopologyProvider instances are shared?

Looking at the `GraphMLMetaTopologyProvider`, everything is cleared in the first part of the reload() function. Which means that anyone accessing the graph while it's being reloaded will end up without any edges/vertices, or existing session may end up in some inconsistent state.